### PR TITLE
alternator/ttl: clean up multi-DC FIXMEs

### LINKED ISSUE #28614: Clean up FIXMEs in alternator/ttl.cc about mutli-DC operation
Currently, Alternator TTL (and after https://github.com/scylladb/scylladb/pull/28320, also per-row TTL in CQL) works as follows on a multi-DC cluster:

1. The expiration-scanning thread only scans data for which this node is the global primary replica. In other words, just one node in the cluster is responsible for checking the expiration of an item - we don't check the same item multiple times on multiple DCs.
2. The expiration-scanning thread reads data using CL=LOCAL_QUORUM. This assumes the data is (eventually) consistent across DCs so we don't need to read with something more drastic like global QUORUM.
3. The deletions are done with CL=LOCAL_QUORUM as well. The writes will eventually be replicated to all DCs, but the scanner doesn't wait for the replication to other DCs (it doesn't use EACH_QUORUM) so in theory may not happen and a cross-DC repair may be needed to complete the replication.

In my opinion, this is a reasonable implementation - we choose performance (no need to repeat the scan on each DC, and also delete expired data in one DC even if a second DC is not available) over absolute consistency. And remember that anyway we have repair to fix consistency, and also, importantly, expiration is _idempotent_ in the sense that if we have a problem with the scan, we can always start it again and delete the items that we forgot to delete.

But when I implemented this, I wasn't sure about this implementation, so `alternator/ttl.cc` is filled with "FIXME" comments second-guessing this decision. The goal of this issue is to remove these pessimistic FIXMEs, perhaps replacing them with one comment explaining the chosen approach.

Although I think the approach we use in the multi-DC case is reasonable, it is not it is also not perfect, so perhaps some specific new FIXMEs or new tests are in order. Here is one:

If a data center is entirely down, it is possible (and perhaps even guaranteed...) that the "secondary" owner of a range is in the same data center, so expiration of some items will not happen while a whole data center is down. The test `test/cluster/test_ttl_row.py::test_row_ttl_multinode_expiration[with_down_node]` from https://github.com/scylladb/scylladb/pull/28320 checks expiration while one node is down, we should do a similar test with an entire data center down.

### PR DESCRIPTION
**Pull request by nyh**
The TTL expiration scanner uses global primary ranges: each token range is scanned by exactly one node in the entire cluster. Reads and deletes use CL=LOCAL_QUORUM; the deletion is replicated to all DCs by normal replication. When a primary node is down, its secondary takes over.

This is a deliberate design choice trading absolute cross-DC consistency for performance (no redundant scans, no waiting for remote DCs). The known limitation is that if an entire DC is down and both the global primary and secondary for a range reside in that DC, expiration of those items is delayed until the DC recovers.

When this code was written, I wasn't sure about this approach and left several FIXMEs second-guessing it. Now that the design is confirmed reasonable (and tested by test_row_ttl_multi_dc), remove those FIXMEs and replace them with clear explanatory comments.

Also: the expire_item() comment was out of date, saying "FIXME!" regarding an Alternator Streams flag for expiration events, which was already implemented with is_system_originated=true. This FIXME is removed and a brief comment explaining this is added.

Finally, add test_row_ttl_multi_dc_with_down_dc to verify the known limitation: some items expire while an entire DC is down (those whose global primary is in a live DC), and all remaining items expire after the DC restarts.

Fixes #28614

### CODE REVIEW

File: alternator/ttl.cc [hunk 1]
```diff
@@ -240,8 +240,8 @@ static bool is_expired(const rjson::value& expiration_time, gc_clock::time_point
 }
 
 // expire_item() expires an item - i.e., deletes it as appropriate for
-// expiration - with CL=QUORUM and (FIXME!) in a way Alternator Streams
-// understands it is an expiration event - not a user-initiated deletion.
+// expiration - with CL=LOCAL_QUORUM, and in a way Alternator Streams
+// understands is an expiration event - not a user-initiated deletion.
 static future<> expire_item(service::storage_proxy& proxy,
```
[the remark this answers is not in the document]
Comment (nyh) [reply alternator/ttl.cc:255]: Good catch.

**Comment by nyh** [comment]
Note to potential reviewers: this is mostly a comment-only patch, plus one new test to confirm the situation described in the new comment.

**Comment by nyh** [comment]
Please review and/or merge. This is a fairly trivial patch.

File: test/cluster/test_ttl_row.py [hunk 2]
```diff
@@ -340,3 +340,88 @@ async def test_row_ttl_multi_dc(manager: ManagerClient):
...
+            # half of them are in primary ranges owned by DC1 so it is almost
+            # certain that some items will expire while DC2 is down.
+            timeout = time.time() + 120
+            while time.time() < timeout:
+                remaining = len(list(await cql_dc1.run_async(
+                    SimpleStatement(f'SELECT p FROM {table}', consistency_level=ConsistencyLevel.LOCAL_QUORUM))))
+                if remaining < 50:
+                    break
+                time.sleep(0.1)
+            remaining_with_dc2_down = len(list(await cql_dc1.run_async(
```
[the remark this answers is not in the document]
Comment (ScyllaPiotr) [reply test/cluster/test_ttl_row.py:431]: I see you handled this.

File: test/cluster/test_ttl_row.py [hunk 3]
```diff
@@ -340,3 +340,88 @@ async def test_row_ttl_multi_dc(manager: ManagerClient):
...
+            # DC2) should also expire.
+            for server in dc2_servers:
+                await manager.server_start(server.server_id)
+
+            timeout = time.time() + 120
+            while time.time() < timeout:
+                if 0 == len(list(await cql.run_async(SimpleStatement(f'SELECT p FROM {table}', consistency_level=ConsistencyLevel.QUORUM)))):
+                    break
+                time.sleep(0.1)
+            assert 0 == len(list(await cql.run_async(SimpleStatement(f'SELECT p FROM {table}', consistency_level=ConsistencyLevel.QUORUM))))
```
[the remark this answers is not in the document]
Comment (ScyllaPiotr) [reply test/cluster/test_ttl_row.py]: I see you handled this.

File: alternator/ttl.cc [hunk 4]
```diff
@@ -161,16 +161,13 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
...
-// when RF=3) need to read the data and send digests.
-// When the expiration thread decides that an item has expired and wants
-// to delete it, it does it using a CL=QUORUM write. This allows this
-// deletion to be visible for consistent (quorum) reads. The deletion,
-// like user deletions, will also appear on the CDC log and therefore
-// Alternator Streams if enabled - currently as ordinary deletes (the
-// userIdentity flag is currently missing this is issue #11523).
+// The expiration thread scans and deletes with CL=LOCAL_QUORUM. Only the
+// global primary owner of each token range performs the scan (see the
+// multi-DC design note near token_ranges_owned_by_this_shard), so each
```
Comment (ScyllaPiotr) [inline alternator/ttl.cc]: I see one at line 386, `token_ranges_owned_by_this_shard` is 60 lines further down.

File: alternator/ttl.cc [hunk 5]
```diff
@@ -161,16 +161,13 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
...
-// locally and in addition QUORUM-1 additional nodes (one additional node
-// when RF=3) need to read the data and send digests.
-// When the expiration thread decides that an item has expired and wants
-// to delete it, it does it using a CL=QUORUM write. This allows this
-// deletion to be visible for consistent (quorum) reads. The deletion,
-// like user deletions, will also appear on the CDC log and therefore
-// Alternator Streams if enabled - currently as ordinary deletes (the
-// userIdentity flag is currently missing this is issue #11523).
+// The expiration thread scans and deletes with CL=LOCAL_QUORUM. Only the
+// global primary owner of each token range performs the scan (see the
```
Comment (ScyllaPiotr) [inline alternator/ttl.cc]: It's not strictly so, sometimes the secondary replica takes over.

File: alternator/ttl.cc [hunk 6]
```diff
@@ -161,16 +161,13 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
...
+// multi-DC design note near token_ranges_owned_by_this_shard), so each
+// item is scanned exactly once per cluster. Deletes are replicated to all
+// DCs by normal replication. Deletions appear on the CDC log and therefore
+// on Alternator Streams as TTL expiration events (with the userIdentity
+// field set), not as ordinary user-initiated deletes.
```
Comment (ScyllaPiotr) [inline alternator/ttl.cc]: They are set... to the value corresponding with `REMOVE`s being `Service`-originated.

File: alternator/ttl.cc [hunk 7]
```diff
@@ -540,9 +547,11 @@ struct scan_ranges_context {
...
         // NOTICE: empty_service_permit is used because the TTL service has fixed parallelism
         query_state_ptr = std::make_unique<service::query_state>(internal_client_state, trace_state, empty_service_permit());
-        // FIXME: What should we do on multi-DC? Will we run the expiration on the same ranges on all
-        // DCs or only once for each range? If the latter, we need to change the CLs in the
-        // scanner and deleter.
+        // Multi-DC: We scan global primary ranges (see comment above on
+        // ranges_holder_primary), so each item is scanned by exactly one node
+        // in the entire cluster. We use LOCAL_QUORUM for both reads and
+        // deletes. Deletions will be replicated to all DCs by the normal
+        // replication mechanism.
```
Comment (ScyllaPiotr) [inline alternator/ttl.cc]: Let's call this "normal replication mechanism" an eventual consistency across DCs, and using "LOCAL_QUORUM" as DC-local strong consistency. Assuming of course I got these right.

File: test/cluster/test_ttl_row.py [hunk 8]
```diff
@@ -337,6 +337,91 @@ async def test_row_ttl_multi_dc(manager: ManagerClient):
...
+       node scans which token range, so each range is scanned by exactly one
+       node across the entire cluster. When DC2 is down, items whose global
+       primary is in DC1 continue to expire normally. Items whose global
+       primary and secondary are both in DC2 will not expire until DC2
+       comes back up - this is a known limitation described in issue #28614.
+       This test verifies that:
+       1. Some items DO expire while DC2 is down (those owned by DC1 primaries).
+       2. After DC2 restarts, all remaining items eventually expire too.
+
+       Note that this test does not enshrine the fact that some items do not
```
Comment (ScyllaPiotr) [inline test/cluster/test_ttl_row.py:367]: I think it would be beneficial if it did enshrine it, so we wouldn't miss it in case it's not our change that caused this. But it's ok as it is now too.

File: test/cluster/test_ttl_row.py [hunk 9]
```diff
@@ -337,6 +337,91 @@ async def test_row_ttl_multi_dc(manager: ManagerClient):
...
+    for dc in range(2):
+        for rack in range(3):
+            futures.append(manager.servers_add(1, config=config, property_file={'dc': f'dc{dc+1}', 'rack': f'rack{rack+1}'}))
+    servers = [x[0] for x in await asyncio.gather(*futures)]
+    # servers[0..2] are in dc1, servers[3..5] are in dc2
+    dc1_server = servers[0]
+    dc2_servers = servers[3:]
+    cql = manager.get_cql()
+    ksdef = "WITH REPLICATION = { 'class': 'NetworkTopologyStrategy', 'replication_factor': 3 }"
+    async with new_test_keyspace(manager, ksdef) as keyspace:
```
Comment (ScyllaPiotr) [inline test/cluster/test_ttl_row.py:389]: The way you spawned a custom ks here made me notice 2 things:
1. We don't have a DynamoDB-compatible way of creating a table with a custom RF class, do we?
2. Did we ever discuss and approve the design for intra-DC TTL behaviour which this PR aims to solidify?

File: test/cluster/test_ttl_row.py [hunk 10]
```diff
@@ -337,6 +337,91 @@ async def test_row_ttl_multi_dc(manager: ManagerClient):
...
+        async with new_test_table(manager, keyspace, 'p int primary key, e bigint') as table:
+            # Enable TTL on column "e" before writing the items.
+            await cql.run_async(f'ALTER TABLE {table} TTL e')
```
Comment (ScyllaPiotr) [inline test/cluster/test_ttl_row.py]: So this test never becomes an Alternator test?

File: test/cluster/test_ttl_row.py [hunk 11]
```diff
@@ -337,6 +337,91 @@ async def test_row_ttl_multi_dc(manager: ManagerClient):
...
+            # With DC2 down, items whose global primary replica is in DC1
+            # should expire promptly. Wait to see at least some expire:
+            # With 50 partitions spread across all token ranges, approximately
+            # half of them are in primary ranges owned by DC1 so it is almost
+            # certain that some items will expire while DC2 is down.
+            timeout = time.time() + 120
+            while time.time() < timeout:
+                remaining = len(list(await cql_dc1.run_async(
+                    SimpleStatement(f'SELECT p FROM {table}', consistency_level=ConsistencyLevel.LOCAL_QUORUM))))
+                if remaining < 50:
```
Comment (ScyllaPiotr) [inline test/cluster/test_ttl_row.py:428]: Let's do a sanity check first that we indeed have 50 items to begin with.

File: test/cluster/test_ttl_row.py [hunk 12]
```diff
@@ -337,6 +337,91 @@ async def test_row_ttl_multi_dc(manager: ManagerClient):
...
+                    break
+                await asyncio.sleep(0.1)
+            remaining_with_dc2_down = len(list(await cql_dc1.run_async(
+                SimpleStatement(f'SELECT p FROM {table}', consistency_level=ConsistencyLevel.LOCAL_QUORUM))))
+            assert remaining_with_dc2_down < 50, "Expected some items to expire with DC2 down"
```
Comment (ScyllaPiotr) [inline test/cluster/test_ttl_row.py:433]: This is insufficient proof. In the edge case, a single TTL-removed item may cause the loop to exit, but the items may still be being expired in DC1, which will later give false hope as to items expiring in DC2 after its restart. The sequence of events I'm describing is improbable, but not impossible. We should have rather:
- emitted an event (and monitored it) on each TTL scan's end
- waited until a single complete scan's run does not reduce the number of items && remaining < 50.

**Review by ScyllaPiotr — CHANGES_REQUESTED** [review]

File: alternator/ttl.cc (hunk 5 above)
Comment (nyh) [reply alternator/ttl.cc]: I'll update the wording to be more accurate. Also, it sort of repeats what was written in paragraphs above. I'll clean it up.

File: alternator/ttl.cc (hunk 6 above)
Comment (nyh) [reply alternator/ttl.cc]: Right, I thought I fixed it, I guess I forgot this place. I'll fix.

Comment (nyh) [reply alternator/ttl.cc]: Wait, what, I did fix it in the text - see - I wrote "Deletions appear on the CDC log and therefore
// on Alternator Streams as TTL expiration events (with the userIdentity
// field set), not as ordinary user-initiated deletes.". This is exactly right.

Comment (nyh) [reply alternator/ttl.cc]: I'm rewriting the entire comment to make it (I hope) clearer.

File: alternator/ttl.cc (hunk 4 above)
Comment (nyh) [reply alternator/ttl.cc]: I'm rewriting the entire comment.

File: alternator/ttl.cc (hunk 7 above)
Comment (nyh) [reply alternator/ttl.cc]: I'll remove this comment completely. No need to have this whole explanation in a comment a second time, and moreover - this comment is inside code that only does reads, not the deletes, so it's not really appropriate to discuss deletes there. We already have a huge comment explaining both scans and deletes elsewhere.

File: test/cluster/test_ttl_row.py (hunk 8 above)
Comment (nyh) [reply test/cluster/test_ttl_row.py:367]: I think it is very important to have a test suite that allows you to refactor code safely, meaning that you expect to most tests to remain useful even when you re-implement or improve some code, and those tests continue to test the new code that it didn't regress anything.

In this case, this is an important test that verifies that even with multiple DCs, "eventually" all items get deleted. No matter how we ever refactor or rewrite the implementation of this feature, this test needs to continue to pass.

If I "enshrine" specific deficiencies that the current implementation has inside this test, it means that if we ever rewrite the implementation without this deficiency, the test will have to be deleted. The author of the new implementation will either outright delete the test or rewrite it, but in any case we'll not know that the regression test continues to pass - because the regression test will have been deleted.

File: test/cluster/test_ttl_row.py (hunk 9 above)
Comment (nyh) [reply test/cluster/test_ttl_row.py:389]: 1. Do you mean https://github.com/scylladb/scylladb/pull/29850? Even with that PR, we can control the replication faction, but not the "class". What other class makes sense besides NetworkTopologyStrategy?
2. No, not really. But this patch doesn't change the implementation, it just changes the comments that claimed it should be fixed (which never happened) , and also tests that the behavior I thought the current implementation has on multiple dcs, really does happen.

File: test/cluster/test_ttl_row.py (hunk 10 above)
Comment (nyh) [reply test/cluster/test_ttl_row.py]: No, I wrote this test for the CQL feature - which uses the same backend TTL scanner as the Alternator feature.

It's not hard to write a parallel test also for Alternator. I didn't think this was very important because I have inside knowledge that exactly the same TTL scanner thread is used in both, so whatever we learn about multi-DC behavior on one, applies to the other as well.

Why did I write this test originally for CQL instead of Alternator? To be honest, I'm not sure :-) I think I was in the mindset that the CQL version of this feature was the newly introduced feature that customers were wishing to use now on multi-DC setups, so I wanted to test that directly.

File: test/cluster/test_ttl_row.py (hunk 11 above)
Comment (nyh) [reply test/cluster/test_ttl_row.py:428]: Ok, I can do that. I'll need to reorder a few statements to do that. Specifically, I must enable the TTL only after finishing writing all 50 items, because if I enable TTL before writing (as I do today), the items may be deleted before I check that 50 exist.

Which reminds me: Maybe I should (optionally?) enable TTL while one DC is down. That's an even more strict test that when the DC comes back up, it joins the TTL work.

File: test/cluster/test_ttl_row.py (hunk 12 above)
Comment (nyh) [reply test/cluster/test_ttl_row.py:433]: But what's wrong with the edge case you describe?
Again, I don't want to enshrine the fact that today while DC2 is down, we don't delete all the items. Maybe in the future we will. The important thing is that:
1. While DC2 is down, DC1 still does some work, doesn't just hang waiting for DC1
2. When DC2 comes back, eventually everything gets expired.

Anyway, I'll change the organization as response to your previous comment, maybe you'll like the new version better.

**Comment by nyh** [comment]
Pushed a new version addressing review comments. Rewrote some of the comments to (hopefully) improve them, and reorganize the test. The test now works in this order:
1. Writes 50 items and verifies they reached both DCs
2. Kill nodes in DC2.
3. Only **now** enable the TTL column. (in the previous version we enabled TTL before killing DC2, so it was possible for the entire TTL to complete before we ever killed DC2).
4. Verify that now that only DC1 is doing expiration work, **some** items get expired (we don't know how many).
5. Bring DC2 back online, and everify that _eventually_ all the items get expired.

File: test/cluster/test_ttl_row.py (hunk 8 above)
Comment (ScyllaPiotr) [reply test/cluster/test_ttl_row.py:367]: > If I "enshrine" specific deficiencies that the current implementation has inside this test, it means that if we ever rewrite the implementation without this deficiency, the test will have to be deleted. The author of the new implementation will either outright delete the test or rewrite it, but in any case we'll not know that the regression test continues to pass - because the regression test will have been deleted.

A short comment about the expected failure of the test in this case should successfully prevent the deletion of the test. But again, we can leave it as it is now too.

File: test/cluster/test_ttl_row.py (hunk 9 above)
Comment (ScyllaPiotr) [reply test/cluster/test_ttl_row.py:389]: 1. That's what we should investigate, not only circumvent in CQL.
2. Yes. And it's good. Perhaps even worth asking GPT to do a post-implementation design doc or a paragraph on this in our user-facing docs.

File: test/cluster/test_ttl_row.py (hunk 10 above)
Comment (ScyllaPiotr) [reply test/cluster/test_ttl_row.py]: I think an Alternator test should exist for it too - IIUC it's not the same implementation, only the same approach (design). So let's guard Alternator's implementation against regression too.

File: test/cluster/test_ttl_row.py (hunk 11 above)
Comment (ScyllaPiotr) [reply test/cluster/test_ttl_row.py:428]: Makes sense!

File: test/cluster/test_ttl_row.py (hunk 12 above)
Comment (ScyllaPiotr) [reply test/cluster/test_ttl_row.py:433]: @copilot I believe the author of the PR misunderstood my intention here, maybe you could explain it better?

Comment (ScyllaPiotr) [reply test/cluster/test_ttl_row.py:433]: ...seems Copilot requires write access to the repo before jumping in like this. Let me copy-paste from another Copilot's window:

"My concern is not primarily that the test must assert that some items remain while DC2 is down. I understand the desire not to enshrine the current limitation if a future implementation can expire everything even with one DC down.

The problem is that the current test does not establish a clean boundary between “items expired by the still-running DC1 work” and “items expired because DC2 came back and resumed its TTL work.”

As soon as the test observes remaining < 50, it restarts DC2. But at that point DC1 may still be in the middle of its own TTL scanning and may continue deleting more items after the restart. If the final remaining == 0 happens after DC2 is restarted, the test may falsely attribute that success to DC2 having resumed scanning, even though those deletions may have been work that DC1 would have completed anyway.

What I wanted is to wait, while DC2 is still down, until the live-DC progress has quiesced — ideally by observing the end of a TTL scan cycle, or at least by waiting until a complete scan/period makes no further reduction. Then:
- if remaining == 0, great: the implementation improved and the test can pass without restarting DC2;
- if 0 < remaining < 50, only then restart DC2 and verify that the remaining rows eventually expire.

So the point is not to permanently require the current limitation. The point is to avoid a false-positive where the post-restart success is actually caused by pre-existing DC1 scanning progress rather than DC2 recovery."

Comment (nyh) [reply test/cluster/test_ttl_row.py:433]: "The problem is that the current test does not establish a clean boundary between “items expired by the still-running DC1 work” and “items expired because DC2 came back and resumed its TTL work.”

Again, I **do not want to** established a boundary between those cases. My entire point is that:
1. I want to make sure DC1 nodes are not idly waiting for DC2 - these nodes have CPUs, they can do some expiration work even when DC2 is down.
2. I don't care, and don't want to enshrine, how much work DC1 is able to do. Currently it can only do (roughly) half the work, but if in the future we'll decide we want to allow it to do more than half the work, that will be fine too.
3. When DC2 comes back up, **eventually** all the items get expired. Again, I didn't want to check maybe all the items get expired even before DC2 comes back up.

You wrote "the point is not to permanently require the current limitation. The point is to avoid a false-positive where the post-restart success is actually caused by pre-existing DC1 scanning progress rather than DC2 recovery.". I don't understand why this is important. If in the future maybe DC1 scanning **will** delete all the items, that will be fine too!

Comment (ScyllaPiotr) [reply test/cluster/test_ttl_row.py:433]: OK, it makes sense.
I re-read the comment before the test and I suggest the following changes:
"This test verifies that:
1. Some items DO expire while DC2 is down (those owned by DC1 primaries).
2. After DC2 restarts, all ~remaining~ items **are** eventually expire**d**. **The expiration may complete before or after starting up of DC2.**"

**Review by ScyllaPiotr — APPROVED** [review]

**Comment by nyh** [comment]
Rebased to fix conflicts and get a fresh CI running, hopefully without those old flaky tests.

File: alternator/ttl.cc [hunk 13]
```diff
@@ -141,37 +141,44 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
...
+// expired items.
+//
+// The expiration-time attribute can be either:
+// 1) a real Scylla column (must be numeric), or
+// 2) an Alternator attribute stored in ATTRS_COLUMN_NAME (map<utf8, bytes>),
+//    where the numeric expiration time is encoded using DynamoDB JSON inside
+//    the bytes value.
+//
+// An expiration thread handles all tables that require expiration scans.
+// Tables are currently scanned sequentially (not in parallel).
```
Comment (nuivall) [inline alternator/ttl.cc]: this is kinda true, but overall not. all shards in the cluster will run expiration_service, so table is globally scanned concurrently, no?

File: test/cluster/test_ttl_row.py [hunk 14]
```diff
@@ -346,6 +346,101 @@ async def test_row_ttl_multi_dc(manager: ManagerClient):
...
+    """
+    config = {
+        'alternator_ttl_period_in_seconds': '0.5',
+    }
+    # Set up a cluster with 6 nodes - three nodes in three racks in each
+    # of two data centers.
+    futures = []
+    for dc in range(2):
+        for rack in range(3):
+            futures.append(manager.servers_add(1, config=config, property_file={'dc': f'dc{dc+1}', 'rack': f'rack{rack+1}'}))
```
Comment (nuivall) [inline test/cluster/test_ttl_row.py:382]: why so odd way to create severs, e.g. we have something like that which should allow any placement, no?
```
    # 4 nodes in 4 racks, default RF=3, therefore not RF-rack-valid
    servers = await manager.servers_add(4, config=config, property_file=[
        {'dc': 'dc1', 'rack': 'rack1'},
        {'dc': 'dc1', 'rack': 'rack2'},
        {'dc': 'dc1', 'rack': 'rack3'},
        {'dc': 'dc1', 'rack': 'rack4'},
    ])
```

**Review by nuivall — APPROVED** [review]

**Comment by nuivall** [comment]
looks good, 2 nits, let me know if you want to fix them before merge

File: alternator/ttl.cc (hunk 13 above)
Comment (nyh) [reply alternator/ttl.cc]: I'll fix. This sentence was meant to qualify the paragraph above which beings with "On each shard, an expiration thread", so it still describes the behavior on each shard. But I agree it's ambiguous.

File: test/cluster/test_ttl_row.py (hunk 14 above)
Comment (nyh) [reply test/cluster/test_ttl_row.py:382]: Why is a loop more or less odd than writing the whole list of 6 nodes manually manually? The way I wrote it already works, so I hesitate changing it...

File: test/cluster/test_ttl_row.py (hunk 11 above)
Comment (nyh) [reply test/cluster/test_ttl_row.py:428]: The code test now checks all 50 items appeared before enabling expiration, and it enables the TTL while DC2 is down. When DC2 is later brought back up, it discovers it needs to do expiration scans and does them, and the test verifies that it does.

File: test/cluster/test_ttl_row.py (hunk 12 above)
Comment (nyh) [reply test/cluster/test_ttl_row.py:433]: "remaining" refers to the items besides those already deleted in step 1. What's wrong with that word?
Did the change from active (expire) to passive (are expired).

The sentence "The expiration may complete before or after starting up of DC2." is already covered by the entire following paragraph, which says exactly the same thing.

File: test/cluster/test_ttl_row.py (hunk 14 above)
Comment (nyh) [reply test/cluster/test_ttl_row.py:382]: Also note that test_row_ttl_multi_dc_with_down_dc is based on the existing test test_row_ttl_multi_dc which also used the same loop to create the cluster, so we'll either need to keep this loop in both, or change it in both, I don't want the new test to use a different idiom than the older test.

**Comment by nyh** [comment]
> looks good, 2 nits, let me know if you want to fix them before merge

I replied to the review comments, fixed some code comments (I hope for the better) and pushed a new version.

File: test/cluster/test_ttl_row.py [hunk 15]
```diff
@@ -346,6 +346,101 @@ async def test_row_ttl_multi_dc(manager: ManagerClient):
...
+            # DC2) should also expire.
+            for server in dc2_servers:
+                await manager.server_start(server.server_id)
+
+            timeout = time.time() + 120
+            while time.time() < timeout:
+                if 0 == len(list(await cql.run_async(SimpleStatement(f'SELECT p FROM {table}', consistency_level=ConsistencyLevel.QUORUM)))):
+                    break
+                await asyncio.sleep(0.1)
+            assert 0 == len(list(await cql.run_async(SimpleStatement(f'SELECT p FROM {table}', consistency_level=ConsistencyLevel.QUORUM))))
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/cluster/test_ttl_row.py]: Good point. I think I had a subconscious (and undocumented) reason to do it the way I did it: Some of the deletions happened on DC1 while DC2 was down, so it's not obvious how these deletions will really reach DC2 during this test. But it should happen - because of hints (or repair, but we don't run repair). Testing hints is not exactly the goal of this test, but arguably hints are part of the reason why it works correctly (deletions already done on DC1 reach DC2 when DC2 comes up) - so should be part of the test.

I'll fix.

**Comment by nyh** [comment]
Pushed a new version addressing a coderabbit suggestion to improve the down-DC test, confirming the items are really deleted from both DCs (requires hints to work correctly, which they do).

**Comment by nyh** [comment]
Pushed a new version updating another outdated comment identified by copilot.

**Comment by swasik** [comment]
@nyh won't it be enough to rebase it and re-trigger ci?
